pubpub/platform
# Wrap server actions in sentry HOF; add custom global-error files

See the [server actions explainer](https://github.com/pubpub/v7/blob/62f50b5ba389fc6e83d4eec2cd12e9d9529ed931/core/lib/serverActions.md) for a description of the server actions-related changes.

This PR:
- Updates Sentry to fix [a very bizarre issue with only certain server actions](https://github.com/getsentry/sentry-javascript/issues/10366).
- Converts deprecated `new Sentry.ReplayIntegration` to `Sentry.replayIntegration()`
- Adds root `global-error.tsx` files to core and integration Next apps. This component should forward all unhandled, **non- server action** application errors to Sentry.
- Introduces `defineServerAction`, which wraps server actions with `Sentry.withServerActionInstrumentation` to properly forward **uncaught errors in server actions** to Sentry. Note, as of now, Sentry's higher-order-function isn't doing much since we're try/catching the result of the server action ourselves.
- Provides a pattern for **caught errors in server actions** to be forwarded to Sentry with `ClientException`...
- Adds a new interface called `ClientException`, which are exceptions sent from server actions to the client when something goes wrong in a server action. These objects should help us standardize how we structure error responses and render them on the client.
- Refactors existing destructive/error toasts to use the newly added `ClientException` interface via a new hook called `useServerAction`. This hook renders the `ClientException` error message and sentry id (if present).
- Adds the Sentry error id to `ClientException` instances where possible, and renders the error id in the exception's subsequent toast.

## Issue(s) Resolved

Resolves #299 

## Test Plan

- Checkout the branch and add some predictable errors in a few server action bodies, and in a couple of routes/pages.
- Comment out the `if (env.NODE_ENV === "production")` check in the `sentry.*.config.ts` files.
- Start the app.
- Trigger the errors.
  - For server actions, you should see a destructive (red) toast when an error occurs. Depending on the server action, you may also see a Sentry error id that looks like the screenshot below.
  - For pages/routes, you should see the default Next error screen.
  - For both, you should see the errors reported in Sentry. Be sure to resolve/close your test errors in Sentry.

## Screenshots (if applicable)

![image](https://github.com/pubpub/v7/assets/6402908/161cc12e-c7c9-4fd8-968a-dd13f70dcdfb)
